Account/FxADeviceRegistrator.swift

```diff
@@ -0,0 +1,125 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Deferred
+import Shared
+
+private let log = Logger.syncLogger
+
+// The current version of the device registration, we use this to re-register
+// devices after we update what we send on device registration.
+private let DeviceRegistrationVersion = 1
+
+public enum FxADeviceRegistrationResult {
+    case RegisteredOrUpdated
+    case AlreadyRegistered
+}
+
+public class FxADeviceRegistrator {
+    public enum DeviceRegistratorError: MaybeErrorType {
+        case DeviceRegistrationFailed
+
+        public var description: String {
+            switch self {
+            case DeviceRegistrationFailed: return "Device registration failed."
+            }
+        }
+    }
+
+    public static func registerOrUpdateDevice(account: FirefoxAccount, state: MarriedState) -> Deferred<Maybe<FxADeviceRegistrationResult>> {
+        if let _ = account.fxaDeviceId where account.deviceRegistrationVersion == DeviceRegistrationVersion {
+            return Deferred(value: Maybe(success: FxADeviceRegistrationResult.AlreadyRegistered))
+        }
+
+        let client = FxAClient10(endpoint: account.configuration.authEndpointURL)
+        let sessionToken = state.sessionToken
+        let name = DeviceInfo.defaultClientName()
+        let device: FxADevice
+        if let deviceId = account.fxaDeviceId {
+            device = FxADevice.forUpdate(name, id: deviceId)
+        } else {
+            device = FxADevice.forRegister(name, type: "mobile")
+        }
+
+        return client.registerOrUpdateDevice(sessionToken, device: device).bind { result in
+            let deferred = Deferred<Maybe<FxADeviceRegistrationResult>>()
```

Comment:
Nit: Instead of defining a deferred variable, you can use the handy `deferMaybe` method to create + fill a deferred:

```swift
return deferMaybe(FXADeviceRegistrationResult.RegisteredOrUpdated)
```